Gcc 4.9

yoyonel opened the pull request:
Some modifications to deal with lib pthread.
Fix somes errors source codes.
Add some flags to deal with C++11 & C++14 (flags are files/sources oriented not projet).
Reactivate some items (or sub-items), with corrections (or not sometimes, maybe gcc-4.9 and flags for C++11&14 are enough ...).

Work & test on Linux Mint 17 (forked of Ubuntu 14.04 LTS), x-64 bits with gcc-4.9.

ps: comments on commits are in french ... sorry :p

ATTYLionelIGN committed:
Suppress .user config from QtCreator IDE

ATTYLionelIGN committed:
Problème de compilation avec certains Items sous GCC-4.9. Desactivation des targets de compilation des items suivants:
=> 16, 23, 27, 31, 32, 33, 34, 35, 36
Il y a des problèmes (en particulier) avec la lib pthread (et/ou CMake)

ATTYLionelIGN committed:
Modification de l'Item13 pour faire tourner les samples C++11 et 14

ATTYLionelIGN committed:
Fix pour faire tourner l'Item16 (pthread). TODO: à revoir, ce n'est pas très propre (trop direct pas modulable/flexible)

ATTYLionelIGN committed:
Items: 23 27 31
Fix pour compilation sous GCC-4.9
Plus de travail pour le 31: Ajout d'une classe singleton pour gérer un pb de doublon d'object dans le header utilitaire 'utils.h'. Ca permet de générer qu'un object 'filters' récupérables maintenant par le singleton associé à une classe dédiée

ATTYLionelIGN committed:
Items: 32 33 36
Fix, réécriture des CMake et partiellement fix de code (36 manquait des points virgules, bizarre)

ATTYLionelIGN committed:
Item34: Semble manquer du code pour cet item. Par exemple la fonction setAlarm ne semble être définie nulle part ....

ATTYLionelIGN committed:
Items: 8 16 21 22 23 27 31 35
Fix et ajout si possible de sous items. Certains codes ou bouts de scripts ne sont pas complets, du coup des fois il n'a pu inclure les examples complètements. Parfois c'est juste des problèmes de CMake et/ou des erreurs d'innattentions dans les sources

ATTYLionelIGN committed:
Fix pour ajouter 'proprement' une gestion de la lib pthread

BartVandewoestyne commented on CMakeLists.txt:
```diff
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 2.6)
 
-project(Effective_Modern_C++ CXX)
+project(Effective_Modern_C++ C CXX)
```
Why exactly must C be added here?  Shouldn't the project be completely CXX?

yoyonel replied:
It's for (cmake) pthread compatibility.
Adding pthread like i do include some tests files (intern codes) to check if pthread is accessible for the current system.
It's a CMake mechanism (througth 'FindThreads.cmake').

MMmMMMM after reading the file 'FindThreads.cmake' (https://github.com/Kitware/CMake/blob/master/Modules/FindThreads.cmake) i'm not sure it's necessary after all ... i will check tomorrow to remove this line.

yoyonel replied:
Oki ... seems strange ... maybe a bug (fix) from CMake.
In my system (from official package 'cmake') in FindThreads.cmake i don't have in the script a test for switching between C and CXX project (like in: https://github.com/Kitware/CMake/blob/master/Modules/FindThreads.cmake around line 94-99).
Maybe the depot (for Ubuntu14.04/Mint) isn't synchronized with the depot of the last release of CMake.
Whatever, with my version (CMake 3.2.2) i need to to add C flag to permit the build of the test file 'CheckForPthreads.c' needed by CMake & 'FindThreads.cmake' to check if a PThread lib is available (with a default/system path).
With the last release of CMake (from github depot), it will not be necessary anymore.

BartVandewoestyne commented on Item13_Prefer_const_iterators_to_iterators/CMakeLists.txt:
```diff
@@ -1,3 +1,8 @@
+add_executable(insertion_cpp11 insertion_cpp11.cpp) # work on gcc-4.9
+
 # TODO add_executable(insertion_cpp11 insertion_cpp14.cpp)
+#set_source_files_properties(insertion_cpp14.cpp PROPERTIES COMPILE_FLAGS -std=c++14)
```
Why is this line commented out?

yoyonel replied:
If it's comment and present here, i think 'insertion_cpp14.cpp' can be compiled without C++14 flag (or support).

yoyonel replied:
And i remember :p
I modified the source 'insertion_cpp14.cpp' and replace cbegin with begin (cend with end) because gcc-4.9 can't deal completely with all C++14 functionnalities (like cbegin, cend).
I've read (somewhere) that gcc-5 it's needed for theses functionnalities. (i will create a new branch 'gcc-5.x' soon ;-))

yoyonel replied:
I've added 'gcc-5.3' branch (https://github.com/yoyonel/Effective-Modern-Cpp/tree/gcc-5.3)
and test the support for std::cbegin and std::cend. 
That's work out_of_the_box (yeaaaahhh ^^). ! (with C++14 flag activated).

BartVandewoestyne commented on Item23_Understand_std_move_and_std_forward/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-set_source_files_properties(std_move_in_cpp14.cpp PROPERTIES COMPILE_FLAGS "-std=c++14 -stdlib=libc++")
+set_source_files_properties(std_move_in_cpp14.cpp PROPERTIES COMPILE_FLAGS "-std=c++14")
```
If I remember correctly, I had to add -stdlib=libc++ to get things to compile with clang.  Will this still compile with clang if you remove this option?

yoyonel replied:
Dunno, but if u want build with GCC (4.9), you have to remove this option (only available on clang compiler (>= 3.5)).
Don't forget, my pull request come from a specific branch 'gcc-4.9' so what it's present on this branch, this is only ("necessary") for gcc-4.9 compiler.
I think a good strategy can be to divided your depot with sub-branches. Each sub-branch will be dedicated for a specific compiler (and version, maybe platform, etc ...).
After this step, the next step can be to merge some sub-branches to get a 'global' compatibility.

yoyonel replied:
Yo Bart ! I've created a new branch 'clang' directly clone from 'gcc-4.9' with some scipts (configure/build) modifications. 
=> https://github.com/yoyonel/Effective-Modern-Cpp/tree/clang-3.5
It seems to build correctly (with no more efforts). 
With my clang-3.5 on my system, the flag '-stdlib=libc++' seems not to be necessary.

BartVandewoestyne replied:
Would it be possible for you to make sure everything builds and compiles with both clang and g++ instead of having separate branches for each of the compiler versions?  I'm quite sure CMake allows you to check what compiler is used and add options depending on compiler.  Feel free to do this and send me the pull request :-)

ATTYLionelIGN committed:
Big refactoring ! (-> not very agile spirit :p)
Add cmake tools inside a dedicated diretory 'cmake/cmake_tools.cmake'.
This tool permit to add subitem(s), deal with flags for C++11 & C++14 compatibilities.
Same for flags/libs needed for pthread and boost.
It's a good start (i think) to unify/merge the manager of multiples builders (gcc, clang, ...).

ATTYLionelIGN committed:
Fix and clean some scripts in CMakeLists.txt

ATTYLionelIGN committed:
remove some 'set_source_files_properties'

ATTYLionelIGN committed:
Add scripts to configure and build with cmake and gcc-4.9.
Fix some a bug with Item22
